File: node/mainNumbers.js
const { Worker } = require('worker_threads');
const config = require('./config');

/**
 * Parallel Batch Processing in Node (mainNumbers)
 * @author Marco Guillén <mguillen.developer@gmail.com>
 */
async function mainNumbers() {
    function worker(numbers) {
        return new Promise((resolve, reject) => {
            const worker = new Worker(config.workerFileNumbers, { workerData: numbers });
            worker.on(config.workerEventMessage, resolve);
            worker.on(config.workerEventError, reject);
        });
    }

    const start = Date.now();
    const numbers = config.numbers
    const segmentsPerWorker = Math.ceil(numbers.length / config.workersForNumbers);
    const batches = Array(config.workersForNumbers).fill().map((_, index) => {
        let images;
        if (index === 0) {
            images = numbers.slice(0, segmentsPerWorker);
        } else if (index === config.workersForNumbers - 1) {
            images = numbers.slice(segmentsPerWorker * index);
        } else {
            images = numbers.slice(segmentsPerWorker * index, segmentsPerWorker * (index + 1))
        }
        return worker(images)
    });
    Promise.all(batches).then(result => console.log(`::: INTENSIVE JOB SUCCESSFULLY COMPLETED -> Workers: ${config.workersForNumbers} -> Elapsed time: ${Date.now() - start} ms -> Total Sum: ${result.reduce((accumulator, currentValue) => accumulator + currentValue)}`)); // Wait for Workers and Promises Group to finish
}

mainNumbers();

File: golang/mainNumbers.go
package main

import (
	"fmt"
	"latamautos/challenge/benchmark/config"
	"math"
	"runtime"
	"sync"
	"time"
)

/**
 * Parallel Batch Processing in Golang (mainImages1)::: INTENSIVE TASKS SUCCESSFULLY COMPLETED -> Elapsed time
 * @author Marco Guillén <mguillen.developer@gmail.com>
 */
func main() {
	runtime.GOMAXPROCS(4)
	var totalSum int64 = 0
	worker := func(numbers []int, wg *sync.WaitGroup) {
		defer wg.Done()
		var sum int64 = 0
		for _, number := range numbers {
			sum += int64(number)
		}
		fmt.Printf("+++ WORKER FINISHED (sum: %d of %d elements)\n", sum, len(numbers))
		totalSum = totalSum + sum
	}

	start := time.Now()
	var wg sync.WaitGroup // Goroutines Group
	segmentsPerWorker := int(math.Ceil(float64(len(config.Config.Numbers)) / float64(config.Config.WorkersForNumbers)))
	for i := 0; i < len(config.Config.Numbers); i += segmentsPerWorker {
		wg.Add(1)
		numbers := config.Config.Numbers[i:config.MinValue(i+segmentsPerWorker, len(config.Config.Numbers))]
		go worker(numbers, &wg)
	}
	wg.Wait() // Wait for Goroutines Group to finish
	duration := time.Since(start)
	fmt.Printf("::: INTENSIVE TASKS SUCCESSFULLY COMPLETED -> Workers: %d -> Elapsed time: %2d ms -> Total Sum: %d) \n", config.Config.WorkersForNumbers, duration.Milliseconds(), totalSum)
}